packages/flutter/lib/src/painting/image_cache.dart

```diff
@@ -247,6 +269,110 @@ class ImageCache {
   }
 }
 
+/// A class that provides information about how the [FlutterImageCache] is
+/// tracking an image.
+///
+/// A [pending] image is one that has not completed yet. It may also be [weak]ly
+/// tracked, but not [strong].
+///
+/// A [strong] image is being held in the LRU cache, and is subject to eviction
+/// based on [ImageCache.maximumSizeBytes] and [ImageCache.maximumSize]. It may
+/// also be [weak]ly tracked, but not [pending].
+///
+/// A [weak] image is being held until its [ImageStreamCompleter] has no more
+/// listeners. It may also be [pending] or [strong].
+/// An [untracked] image is not being cached.
+///
+/// To obtain an [ImageCacheLocation], use [FlutterImageCache.locationForKey] or
+/// [ImageProvider.findCacheLocation].
+class ImageCacheLocation {
+  const ImageCacheLocation._({
+    this.pending = false,
+    this.strong = false,
+    this.weak = false,
+  }) : assert(!pending || !strong);
+
+  /// An image that has been submitted to [FlutterImageCache.putIfAbsent], but
+  /// not yet completed.
+  final bool pending;
+
+  /// An image that has been submitted to [FlutterImageCache.putIfAbsent], has
+  /// completed, fits based on the sizing rules of the cache, and has not been
+  /// evicted.
+  final bool strong;
+
+  /// An image that has been submitted to [FlutterImageCache.putIfAbsent], has
+  /// completed, and has at least one listener on its [ImageStreamCompleter].
+  final bool weak;
+
+  /// An image that either has not been submitted to
+  /// [FlutterImageCache.putIfAbsent] or has otherwise been evicted from the
+  /// [strong] and [weak] caches.
+  bool get untracked => !pending && !strong && !weak;
+
+  @override
+  bool operator ==(Object other) {
+    if (other.runtimeType != runtimeType) {
+      return false;
+    }
+    return other is ImageCacheLocation
+        && other.pending == pending
+        && other.strong == strong
+        && other.weak == weak;
+  }
+
+  @override
+  int get hashCode => hashValues(pending, strong, weak);
+}
+
+/// An [ImageCache] that adds weak tracking of images that have at least one
+/// listener on their [ImageStreamCompleter].
+///
+/// This cache also exposes internal dimensions of the implementation, such as
+/// checking whether an image for a given key is pending, weak, or completed.
+/// See [ImageCacheLocation] for more details.
+class FlutterImageCache extends ImageCache {
```

Comment:
in the api design review meeting we discussed just merging FlutterImageCache into ImageCache, and make the test-only features be methods called `debugFoo...` with their bodies in asserts.

Reply:
I'm leaning towards making these non-debug/assert methods at this point - this could be valuable information at runtime (e.g. if an application wants to know how many live images are on the screen before decoding more).  I'm also documenting [precacheImage] as relying on this behavior now.